File: elixir/rail-fence-cipher/rail_fence_cipher.exs
defmodule RailFenceCipher do
  @doc """
  Encode a given plaintext to the corresponding rail fence ciphertext
  """
  @spec encode(String.t, pos_integer) :: String.t
  def encode(str, rails) do
    graphemes = String.graphemes(str)
    cond do
      rails > 1 -> stream = Enum.to_list(0..(rails-1)) ++ Enum.to_list((rails-2)..1)
      rails == 1 -> stream = [0]
    end
    bar = Stream.cycle(stream) |> Enum.take(length(graphemes))
    chars = Enum.zip(graphemes, bar)
            |> Enum.group_by(fn {_, i} -> i end)
    for {_, rail} <- chars, {c, _} <- rail do
      c
    end
    |> Enum.join
  end

  @doc """
  Decode a given rail fence ciphertext to the corresponding plaintext
  """
  @spec decode(String.t, pos_integer) :: String.t
  def decode(str, rails) do
    graphemes = String.graphemes(str)
    cond do
      rails > 1 -> stream = Enum.to_list(0..(rails-1)) ++ Enum.to_list((rails-2)..1)
      rails == 1 -> stream = [0]
    end
    bar =
      stream
      |> Stream.cycle
      |> Enum.take(length graphemes)
      |> Enum.zip(0..(length(graphemes)))
      |> Enum.sort
      |> Enum.zip(graphemes)
      |> Enum.sort(fn {{_, a}, _}, {{_, b}, _} -> a < b end)

    (for {{_, _}, c} <- bar, do: c)
    |> Enum.join
  end
end

File: python/rail-fence-cipher/rail_fence_cipher.py
from itertools import cycle


def fence_pattern(message, rails):
    rails_indexes = cycle(list(range(rails)) + list(range(rails-2, 0, -1)))
    indexed_chars = zip(rails_indexes, message)
    fence = [[] for _ in range(rails)]
    for (i, c) in indexed_chars:
        fence[i].append(c)
    return fence


def encode(message, rails):
    fence = fence_pattern(message, rails)
    return ''.join([c for rail in fence for c in rail])


def decode(encoded_message, rails):
    msg_len = len(encoded_message)
    indexes = fence_pattern(range(msg_len), rails)
    indexes = [n for row in indexes for n in row]  # flatten
    indexes = zip(indexes, range(msg_len))
    read_sequence = [i for (_, i) in sorted(indexes)]
    return ''.join([encoded_message[i] for i in read_sequence])
